### ISSUE TITLE
Questions regarding performance

### ISSUE THREAD

**Issue by Mirorrn** [StanfordASL/Trajectron-plus-plus#37]
Hello,

first of all many thanks for publishing the code!

I have a question regarding the result in table 2 in the paper. What difference to the previous approach "Trajectron" has brought this huge prediction increase? As far as I understand there is no difference between the architectures. I see that with "Trajectron++" you have chosen the Info-VAE approach as well as a discrete z variable. 

My second question is about the evaluation. In the "Social-GAN", each sample consists of the prediction of the whole scene. A total of 20 samples are created and then the best scene is chosen based on the ADE. Do you evaluate each trajectory separately or also based on the whole scene?

Thanks in advance :)

**Comment by BorisIvanovic** [comment]
Hi @Mirorrn,

1. You're right that Trajectron and the stripped-down baseline version of Trajectron++ are similar. However, we have made changes to internal components of the architecture while still keeping the overall idea the same. For example, here's one of the changes we made: 
    - Trajectron does the typical Gumbel-Softmax reparametrized sampling to get z ~ q(z | x, y) during training.
    - Trajectron++ does not do this, instead just enumerating all z values and explicitly computing the log_likelihood of the GT data using the full GMM distribution. This is also mentioned in the paper, in the paragraph below the loss function (Eq. 4)

2. I don't know if I agree with that assessment of what S-GAN does for evaluation. Taking a look at their released code, we see that the authors chose to select the best ADE and FDE values independent of each other (meaning it is not actually the best scene that is chosen by a specific metric, but rather the best separate trajectories are chosen per metric): https://github.com/agrimgupta92/sgan/blob/master/scripts/evaluate_model.py#L84 Like them, we evaluate each trajectory separately.

**Comment by Mirorrn** [comment]
1. thank you for the clarification!

2. S-GAN basic idea is to map the multimodality of the whole scene, therefore the whole scene is evaluated. The following line shows how this is done: https://github.com/agrimgupta92/sgan/blob/691231e1adb6a344c7bcea9ebf2534518b226ead/scripts/evaluate_model.py#L44
The input here is a matrix, of the form error=[#sample,#number of traj. in batch]. After stacking: error = torch.stack(error, dim=1) is error=[#number of traj. in batch,#sample]. In the for-loop, the error of the trajectories of a scene with "start, end" is selected from the serialized batch and then summed up. Based on the sum of the errors per scene, the scene sample with the lowest error is selected.

I don't want to expose anything, just mention that the other approaches evaluate differently, which shows the huge gap in the evaluation.

**Comment by BorisIvanovic** [comment]
No problem!

1. Happy to help!
2. I think I see what you mean, although I still find it a bit weird to do that for a model with such randomness (again, S-GAN does not generate scene-consistent predictions).

    I think the correct answer here is to stop using minFDE/minADE metrics. I always disliked them (and am not a fan when reviewers ask for them) ever since our ICCV 2019 Trajectron paper when we proposed the KDE-based NLL metric as a substitute. We use this KDE-based NLL metric in Table 3 of the Trajectron++ and I think that is the better comparison point for these models.

3. The only difference between the last row of (a) and the first row of (b) is that in (b) the ego-vehicle is _not_ included as an agent-to-be-evaluated at test-time. We did this because the ego-vehicle's future is being conditioned on in the last row of (b) and we wanted to show that any resulting aggregate performance improvements were not caused solely because of the removal of the ego-vehicle during evaluation (since it's being conditioned on).

    Good question! We did indeed just take a random pedestrian and consider it as the robot in a scene, and that is how we determined y_R for the training of the robot-inclusive Trajectron++ there.

**Comment by Mirorrn** [comment]
Thank you very much now I think everything is so far clarified!
